README.md
# exegesis-plugin-jsonschema

[![Run Status](https://api.shippable.com/projects/5c437f657433f60600a12e74/badge?branch=master)]()
[![Coverage Badge](https://api.shippable.com/projects/5c437f657433f60600a12e74/coverageBadge?branch=master)]()
![](https://img.shields.io/github/issues/phil-mitchell/exegesis-plugin-jsonschema.svg)
![](https://img.shields.io/github/license/phil-mitchell/exegesis-plugin-jsonschema.svg)
![](https://img.shields.io/node/v/exegesis-plugin-jsonschema.svg)

## Description

An exegesis plugin to generate CRUD operations based on JSON schema definition.

## Installation

```sh
npm install exegesis-plugin-jsonschema
```

## Example

Add this to your Exegesis options:

```js
const exegesisJSONSchemaPlugin = require( 'exegesis-plugin-jsonschema' );

var jsonShemaPlugin = exegesisJSONSchemaPlugin({});
await jsonSchemaPlugin.addSchema( path.resolve( __dirname, './model/root.json' ) );

options = {
    plugins: [
        jsonSchemaPlugin
    ]
};
```


CHANGELOG.md
## [1.2.3](https://github.com/phil-mitchell/exegesis-plugin-jsonschema/compare/v1.2.2...v1.2.3) (2021-05-09)


### Bug Fixes

* **deps:** Update dependencies to fix security warnings ([162baac](https://github.com/phil-mitchell/exegesis-plugin-jsonschema/commit/162baacf48763421a21616cfa1b00525336a4ead))

## [1.2.2](https://github.com/phil-mitchell/exegesis-plugin-jsonschema/compare/v1.2.1...v1.2.2) (2021-05-09)


### Bug Fixes

* **build:** set up PR testing ([af5b083](https://github.com/phil-mitchell/exegesis-plugin-jsonschema/commit/af5b083f77f00a015490923f755ae2a2e3163455))

## [1.2.1](https://github.com/phil-mitchell/exegesis-plugin-jsonschema/compare/v1.2.0...v1.2.1) (2021-01-24)


### Bug Fixes

* **extraProperties:** handle schema with extraProperties specified ([#11](https://github.com/phil-mitchell/exegesis-plugin-jsonschema/issues/11)) ([7271255](https://github.com/phil-mitchell/exegesis-plugin-jsonschema/commit/7271255a02ff51e3fa861e0ab21955423716933c))

# [1.2.0](https://github.com/phil-mitchell/exegesis-plugin-jsonschema/compare/v1.1.4...v1.2.0) (2020-11-23)


### Features

* **operations:** Support operations defined in the jsonschema ([7606c45](https://github.com/phil-mitchell/exegesis-plugin-jsonschema/commit/7606c452fbc1e604e63720ee281696409c1c1b92))

## [1.1.4](https://github.com/phil-mitchell/exegesis-plugin-jsonschema/compare/v1.1.3...v1.1.4) (2020-11-22)


### Bug Fixes

* **id:** Specify the id field and use its schema for the path parameter ([cc02fe8](https://github.com/phil-mitchell/exegesis-plugin-jsonschema/commit/cc02fe8d6bc7ba2aae6fa5609340ff0a5f0a8290))

## [1.1.3](https://github.com/phil-mitchell/exegesis-plugin-jsonschema/compare/v1.1.2...v1.1.3) (2020-02-01)


### Bug Fixes

* **config:** update pipeline to use Node 10 ([89adfeb](https://github.com/phil-mitchell/exegesis-plugin-jsonschema/commit/89adfebd0a1936c17db6480a456835ff994bb892))

## [1.1.2](https://github.com/phil-mitchell/exegesis-plugin-jsonschema/compare/v1.1.1...v1.1.2) (2019-08-17)


### Bug Fixes

* **dependencies:** update dependencies ([47704d5](https://github.com/phil-mitchell/exegesis-plugin-jsonschema/commit/47704d5))

## [1.1.1](https://github.com/phil-mitchell/exegesis-plugin-jsonschema/compare/v1.1.0...v1.1.1) (2019-04-15)


### Bug Fixes

* **api:** apply correct defaults if default is falsey ([2d9db2f](https://github.com/phil-mitchell/exegesis-plugin-jsonschema/commit/2d9db2f))

# [1.1.0](https://github.com/phil-mitchell/exegesis-plugin-jsonschema/compare/v1.0.2...v1.1.0) (2019-03-09)


### Features

* **api:** support models in arbitrary acyclic graphs ([8f3bb63](https://github.com/phil-mitchell/exegesis-plugin-jsonschema/commit/8f3bb63))

## [1.0.2](https://github.com/phil-mitchell/exegesis-plugin-jsonschema/compare/v1.0.1...v1.0.2) (2019-02-11)


### Bug Fixes

* **api:** add base path to the operation object ([9f6937f](https://github.com/phil-mitchell/exegesis-plugin-jsonschema/commit/9f6937f))

## [1.0.1](https://github.com/phil-mitchell/exegesis-plugin-jsonschema/compare/v1.0.0...v1.0.1) (2019-02-09)


### Bug Fixes

* **api:** use uniqe names for operationIds to work better with swagger-ui ([abaeb38](https://github.com/phil-mitchell/exegesis-plugin-jsonschema/commit/abaeb38))

# 1.0.0 (2019-01-27)


### Bug Fixes

* **release:** add correct badges for shippable ([9c694fd](https://github.com/phil-mitchell/exegesis-plugin-jsonschema/commit/9c694fd))


test/app.js
'use strict';

const express = require( 'express' );
const exegesisExpress = require( 'exegesis-express' );
const http = require( 'http' );
const path = require( 'path' );
const fs = require( 'fs-extra' );

const require_helper = require( './require_helper' );
const JSONSchemaPlugin = require_helper( 'index.js' );

async function createServer() {
    const app = express();

    var jsonSchemaPlugin = JSONSchemaPlugin({
        schema: {
            baseUrl: '/constructor',
            schema: {
                properties: {
                    employees: require( './models/basic/employee.json' )
                }
            }
        }
    });
    await jsonSchemaPlugin.addSchema( __dirname + '/models/basic/root.json', '/basic' );
    await jsonSchemaPlugin.addSchema( __dirname + '/models/branching/root.json', '/branching' );

    const options = {
        controllers: require( './controllers' ),
        allowMissingControllers: false,
        plugins: [ jsonSchemaPlugin ]
    };

    const exegesisMiddleware = await exegesisExpress.middleware(
        path.resolve( __dirname, './petstore.yaml' ),
        options
    );

    app.use( exegesisMiddleware );

    // Return a 404
    app.use( ( req, res ) => {
        res.status( 404 ).json({ message: `Not found` });
    });

    // Handle any unexpected errors
    app.use( ( err, req, res, next ) => {
        res.status( 500 ).json({ message: `Internal error: ${err.message}` });
        next();
    });

    const server = http.createServer( app );

    return server;
}

if( process.env.REPORT_DIR_FOR_CODE_COVERAGE ) {
    const dumpCoverage = () => {
        console.warn( 'Outputting code coverage information to ' + process.env.REPORT_DIR_FOR_CODE_COVERAGE );
        fs.ensureDirSync( process.env.REPORT_DIR_FOR_CODE_COVERAGE );
        fs.writeFileSync(
            process.env.REPORT_DIR_FOR_CODE_COVERAGE + '/app.json',
            JSON.stringify( global['__coverage__'] ), 'utf8'
        );
    };
    process.on( 'exit', dumpCoverage );
    process.on( 'SIGINT', process.exit );
    process.on( 'SIGTERM', process.exit );
}

createServer()
.then( server => {
    server.listen( 3000 );
    console.log( 'Listening on port 3000' );
    console.log( 'Try visiting http://localhost:3000/api-docs' );
})
.catch( err => {
    console.error( err.stack );
    process.exit( 1 );
});


test/unit/index.spec.js
'use strict';

var expect = require( 'chai' ).expect;
var require_helper = require( '../require_helper' );
var index = require_helper( 'index' );

describe( 'Schema option', function() {
    it( 'creates a default schema as an empty array', function() {
        var plugin = index({});
        expect( plugin.options.schema ).to.eql( [] );
    });

    it( 'creates an array if schema is not an array', function() {
        var plugin = index({ schema: {} });
        expect( plugin.options.schema ).to.eql( [ {} ] );
    });

    it( 'can add a new schema to the list with no baseUrl', async function() {
        var plugin = index();
        expect( plugin.options.schema ).to.eql( [] );
        await plugin.addSchema({});
        expect( plugin.options.schema ).to.eql( [ { baseUrl: '', schema: {} } ] );
    });

    it( 'can add a new schema to the list with a baseUrl', async function() {
        var plugin = index();
        expect( plugin.options.schema ).to.eql( [] );
        await plugin.addSchema({}, '/test/api/base' );
        expect( plugin.options.schema ).to.eql( [ { baseUrl: '/test/api/base', schema: {} } ] );
    });
});

describe( 'JSONSchemaPlugin', function() {
    describe( 'openapi version check', function() {
        before( function() {
            this.plugin = index({});
        });
        it( 'throws an error if there is no openapi', function() {
            expect( this.plugin.makeExegesisPlugin.bind(
                this.plugin, {
                    apiDoc: {
                    }
                }) ).to.throw( "OpenAPI definition is missing 'openapi' field" );
        });
        it( 'throws an error if openapi is too low', function() {
            expect( this.plugin.makeExegesisPlugin.bind(
                this.plugin, {
                    apiDoc: {
                        openapi: '2.0.0'
                    }
                }) ).to.throw( 'OpenAPI version 2.0.0 not supported' );
        });
        it( 'throws an error if openapi is too high', function() {
            expect( this.plugin.makeExegesisPlugin.bind(
                this.plugin, {
                    apiDoc: {
                        openapi: '4.0.0'
                    }
                }) ).to.throw( 'OpenAPI version 4.0.0 not supported' );
        });
    });

    describe( 'preCompile', function() {
        before( function() {
            this.plugin = index({});
        });
        it( 'adds controllers when options have no controllers included', function() {
            var instance = this.plugin.makeExegesisPlugin({
                apiDoc: {
                    openapi: '3.0.0'
                }
            });
            var params = {
                api: {},
                options: {}
            };
            instance.preCompile( params );
            expect( params.options ).to.have.property( 'controllers' );
            expect( params.options.controllers ).to.have.property( instance.controller.name );
            expect( params.options.controllers[instance.controller.name] ).to.eql( instance.controller );
            expect( instance.controller._controllers ).to.eql( params.options.controllers );
        });
        it( 'adds controllers to existing set of controllers', function() {
            var instance = this.plugin.makeExegesisPlugin({
                apiDoc: {
                    openapi: '3.0.0'
                }
            });
            var params = {
                api: {},
                options: {
                    controllers: {
                        testController: {}
                    }
                }
            };
            instance.preCompile( params );
            expect( params.options ).to.have.property( 'controllers' );
            expect( params.options.controllers ).to.have.property( instance.controller.name );
            expect( params.options.controllers[instance.controller.name] ).to.eql( instance.controller );
            expect( instance.controller._controllers ).to.eql( params.options.controllers );
            expect( params.options.controllers ).to.have.property( 'testController' );
            expect( params.options.controllers.testController ).to.eql({});
        });
        it( 'does not overrite existing controllers', function() {
            var instance = this.plugin.makeExegesisPlugin({
                apiDoc: {
                    openapi: '3.0.0'
                }
            });
            var params = {
                api: {},
                options: {
                    controllers: {}
                }
            };
            params.options.controllers[instance.controller.name] = {};

            expect( instance.preCompile.bind( instance, params ) ).to.throw( 'Controller already added' );

            expect( params.options ).to.have.property( 'controllers' );
            expect( params.options.controllers ).to.have.property( instance.controller.name );
            expect( params.options.controllers[instance.controller.name] ).to.not.eql( instance.controller );
            expect( instance.controller._controllers ).to.not.eql( params.options.controllers );
        });
    });
});


test/controllers/index.js
'use strict';

module.exports.employee = require( './employee' );


src/index.js
'use strict';
const semver = require( 'semver' );
const parser = require( 'json-schema-ref-parser' );

const JSONSchemaController = require( './controller' );
const pathGenerator = require( './pathGenerator' );

class JSONSchemaPlugin {
    constructor( apiDoc, options ) {
        this._options = options;

        // Verify the apiDoc is an OpenAPI 3.x.x document, because this plugin
        // doesn't know how to handle anything else.
        if( !apiDoc.openapi ) {
            throw new Error( "OpenAPI definition is missing 'openapi' field" );
        }

        if( !semver.satisfies( apiDoc.openapi, '>=3.0.0 <4.0.0' ) ) {
            throw new Error( `OpenAPI version ${apiDoc.openapi} not supported` );
        }

        this.controller = new Proxy( new JSONSchemaController(), {
            has: function( object, property ) {
                if( typeof( property ) !== 'string' || Reflect.has( object, property ) ) {
                    return true;
                }
                return Reflect.has( object, property.split( ' ' )[0] );
            },
            get: function( object, property ) {
                if( typeof( property ) !== 'string' || Reflect.has( object, property ) ) {
                    return Reflect.get( object, property );
                }
                return Reflect.get( object, property.split( ' ' )[0] );
            }
        });

        this._options.schema.forEach( schema => {
            pathGenerator( apiDoc, this.controller.name, schema.schema, schema.baseUrl );
        });
    }

    // Called exactly once, before Exegesis "compiles" the API document.
    // Plugins must not modify apiDoc here.
    preCompile({ options }) {
        options.controllers = options.controllers || {};
        if( options.controllers.hasOwnProperty( this.controller.name ) ) {
            throw new Error( `Controller already added with name ${this.controller.name}` );
        }
        this.controller.controllers = options.controllers;
        options.controllers[this.controller.name] = this.controller;
    }
}

module.exports = function plugin( options ) {
    options = options || {};
    if( !options.schema ) {
        options.schema = [];
    }
    if( !Array.isArray( options.schema ) ) {
        options.schema = [ options.schema ];
    }
    return{
        info: {
            // This should match the name of your npm package.
            name: 'exegesis-plugin-jsonschema'
        },
        options: options,
        makeExegesisPlugin: function makeExegesisPlugin({ apiDoc }) {
            return new JSONSchemaPlugin( apiDoc, this.options );
        },
        addSchema: async function addSchema( schema, baseUrl ) {
            this.options.schema.push({
                baseUrl: baseUrl || '',
                schema: await parser.dereference( schema )
            });
        }
    };
};


src/blankObject.js
'use strict';

const schemaWalker = require( '@cloudflare/json-schema-walker' );

module.exports = function blankObject( schemaRoot ) {
    function preSchema( schema, path, parent ) {
        if( typeof( schema ) !== 'object' ) {
            return schema;
        }
        switch( schema.type ) {
        case'array':
            schema.res = schema.res || [];
            break;
        case'object':
            schema.res = schema.res || {};
            break;
        default:
            schema.res = schema.res || ( schema.properties && {}) || null;
        }

        if( ( schema.type === 'array' && schema.items && schema.items.$id ) ||
            ( parent && ( schema.type === 'object' || schema.properties ) && schema.$id ) ||
            schema.readOnly ) {
            delete schema.items;
            delete schema.properties;
            schema.type = 'string';
            schema.format = 'uri';
            schema.readOnly = true;
        } else if( schema.default !== undefined ) {
            schema.res = schema.default;
        }

        if( schema.type === 'array' ) {
            schema.type = 'string';
            schema.readOnly = true;
            delete schema.items;
        }
        return schema;
    }

    function postSchema( schema, path, parent ) {
        if( typeof( schema ) !== 'object' ) {
            return;
        }
        if( parent && !schema.readOnly && parent.type === 'object' ) {
            var res = null;
            if( schema.res != null ) {
                res = schema.res;
            } else if( parent.res[path[path.length-1]] != null ) {
                res = parent.res[path[path.length-1]];
            }
            parent.res[path[path.length-1]] = res;
        }
    }

    schemaRoot = JSON.parse( JSON.stringify( schemaRoot ) );
    const vocab = schemaWalker.getVocabulary( schemaRoot, schemaWalker.vocabularies.DRAFT_04 );
    schemaWalker.schemaWalk( schemaRoot, preSchema, postSchema, vocab );
    return schemaRoot.res;
};


src/pathGenerator.js
'use strict';

const blankObject = require( './blankObject' );
const toOpenAPIComponent = require( './toOpenAPIComponent' );

class JSONSchemaPathGenerator {
    constructor( apiDoc, controller, baseUrl ) {
        if( typeof( apiDoc ) !== 'object' ) {
            throw new Error( 'apiDoc object is required' );
        }
        if( typeof( controller ) !== 'string' ) {
            throw new Error( 'controller name is required' );
        }

        apiDoc.paths = apiDoc.paths || {};
        apiDoc.components = apiDoc.components || {};
        apiDoc.components.schemas = apiDoc.components.schemas || {};

        this.apiDoc = apiDoc;
        this.controller = controller;

        this.namespace = ( baseUrl || '' ).replace( /\//g, '.' ) || '';
        this.baseUrl = baseUrl || '';
    }

    toSchemaName( context, schema ) {
        if( !schema.title ) {
            throw new Error( `Missing title in schema: ${ JSON.stringify( schema ) }` );
        }
        if( this.namespace ) {
            return`${this.namespace}::${schema.title}`;
        }
        return schema.title;
    }

    addObjectDefinition( context, schema ) {
        var title = this.toSchemaName( context, schema );
        var apiSchema = toOpenAPIComponent( schema );
        if( this.apiDoc.components.schemas[title] ) {
            if( JSON.stringify( apiSchema ) !== JSON.stringify( this.apiDoc.components.schemas[title] ) ) {
                throw new Error( `Duplicate schema with title ${title}` );
            }
        } else {
            this.apiDoc.components.schemas[title] = apiSchema;
        }
    }

    addObjectPath( context, schema ) {
        var localControllerName = schema['exegesis-plugin-jsonschema-controller'];
        if( !localControllerName ) {
            throw new Error( `Schema ${schema.title} is missing exegesis-plugin-jsonschema-controller` );
        }

        var urlPath = ( ( this.baseUrl || '' ) + ( context.urlPath || '' ) ) || '/';
        if( this.apiDoc.paths[urlPath] ) {
            throw new Error( `Duplicate path ${urlPath || '/'} found for schema ${schema.title}` );
        }
        var schemaName = this.toSchemaName( context, schema );

        this.apiDoc.paths[urlPath] = {
            'x-exegesis-controller': this.controller,
            'x-exegesis-jsonschema-controller': localControllerName,
            'x-exegesis-jsonschema-blankobject': JSON.stringify( blankObject( schema ) ),
            'x-exegesis-jsonschema-pathbase': this.baseUrl || '',
            'x-exegesis-jsonschema-pathtemplate': context.urlPath || '/',
            get: {
                summary: `Gets a single ${schema.title || 'item'}`,
                operationId: `getItem ${urlPath}`,
                parameters: [].concat( context.pathParams || [] ),
                responses: {
                    '200': {
                        description: `The ${schema.title}`,
                        content: {
                            'application/json': {
                                schema: {
                                    $ref: `#/components/schemas/${schemaName}`
                                }
                            }
                        }
                    }
                }
            },
            patch: {
                summary: `Modify a ${schema.title}`,
                operationId: `patchItem ${urlPath}`,
                parameters: [].concat( context.pathParams || [] ),
                requestBody: {
                    description: `The modified ${schema.title} values`,
                    required: true,
                    content: {
                        'application/json': {
                            schema: {
                                $ref: `#/components/schemas/${schemaName}`
                            }
                        }
                    }
                },
                responses: {
                    '200': {
                        description: `The modified ${schema.title}`,
                        content: {
                            'application/json': {
                                schema: {
                                    $ref: `#/components/schemas/${schemaName}`
                                }
                            }
                        }
                    }
                }
            },
            put: {
                summary: `Replace a ${schema.title}`,
                operationId: `putItem ${urlPath}`,
                parameters: [].concat( context.pathParams || [] ),
                requestBody: {
                    description: `The new ${schema.title} to replace this one`,
                    required: true,
                    content: {
                        'application/json': {
                            schema: {
                                $ref: `#/components/schemas/${schemaName}`
                            }
                        }
                    }
                },
                responses: {
                    '200': {
                        description: `The replaced ${schema.title}`,
                        content: {
                            'application/json': {
                                schema: {
                                    $ref: `#/components/schemas/${schemaName}`
                                }
                            }
                        }
                    }
                }
            }
        };

        if( !schema.readOnly ) {
            this.apiDoc.paths[urlPath]['delete'] = {
                summary: `Delete a ${schema.title}`,
                operationId: `deleteItem ${urlPath}`,
                parameters: [].concat( context.pathParams || [] ),
                responses: {
                    '204': {
                        description: `No content`
                    }
                }
            };
        }
    }

    addCollectionPath( context, schema ) {
        var localControllerName = schema['exegesis-plugin-jsonschema-controller'];
        if( !localControllerName ) {
            throw new Error( `Schema ${schema.title} is missing exegesis-plugin-jsonschema-controller` );
        }

        var urlPath = ( ( this.baseUrl || '' ) + ( context.urlPath || '' ) ) || '/';
        if( this.apiDoc.paths[urlPath || '/'] ) {
            throw new Error( `Duplicate path ${urlPath || '/'} found for schema ${schema.title}` );
        }
        var schemaName = this.toSchemaName( context, schema );

        this.apiDoc.paths[urlPath] = {
            'x-exegesis-controller': this.controller,
            'x-exegesis-jsonschema-controller': localControllerName,
            'x-exegesis-jsonschema-blankobject': JSON.stringify( blankObject( schema ) ),
            'x-exegesis-jsonschema-pathbase': this.baseUrl || '',
            'x-exegesis-jsonschema-pathtemplate': context.urlPath || '/',
            get: {
                summary: `Gets a list of ${schema.title}`,
                operationId: `getItems ${urlPath}`,
                parameters: [].concat( context.pathParams || [] ),
                responses: {
                    '200': {
                        description: `The list of ${schema.title}`,
                        content: {
                            'application/json': {
                                schema: {
                                    type: 'array',
                                    items: {
                                        $ref: `#/components/schemas/${schemaName}`
                                    }
                                }
                            }
                        }
                    }
                }
            }
        };
        if( !schema.readOnly ) {
            this.apiDoc.paths[urlPath].post = {
                summary: `Create a new ${schema.title}`,
                operationId: `postItems ${urlPath}`,
                parameters: [].concat( context.pathParams || [] ),
                requestBody: {
                    description: `The new ${schema.title} to be created`,
                    required: true,
                    content: {
                        'application/json': {
                            schema: {
                                $ref: `#/components/schemas/${schemaName}`
                            }
                        }
                    }
                },
                responses: {
                    '201': {
                        description: `The new ${schema.title}`,
                        content: {
                            'application/json': {
                                schema: {
                                    $ref: `#/components/schemas/${schemaName}`
                                }
                            }
                        }
                    }
                }
            };
        }
    }

    addOperationPath( context, schema, operation ) {
        var localControllerName = schema['exegesis-plugin-jsonschema-controller'];
        if( !localControllerName ) {
            throw new Error( `Schema ${schema.title} is missing exegesis-plugin-jsonschema-controller` );
        }

        var urlPath = ( ( this.baseUrl || '' ) + ( context.urlPath || '' ) ) || '/';
        if( this.apiDoc.paths[urlPath || '/'] ) {
            throw new Error( `Duplicate path ${urlPath || '/'} found for schema ${schema.title}` );
        }

        this.apiDoc.paths[urlPath] = {
            'x-exegesis-controller': this.controller,
            'x-exegesis-jsonschema-controller': localControllerName,
            'x-exegesis-jsonschema-blankobject': JSON.stringify( blankObject( operation.schema ) ),
            'x-exegesis-jsonschema-pathbase': this.baseUrl || '',
            'x-exegesis-jsonschema-pathtemplate': context.urlPath || '/',
            post: {
                summary: operation.summary,
                operationId: `callOperation ${urlPath}`,
                parameters: [].concat( context.pathParams || [] ),
                requestBody: {
                    description: `The operation parameters`,
                    required: true,
                    content: {
                        'application/json': {
                            schema: operation.schema
                        }
                    }
                },
                responses: operation.responses
            }
        };
    }

    processSchema( context, schema, collection, skipPath ) {
        if( collection ) {
            if( !skipPath ) {
                this.addCollectionPath( context, schema );
            }
            let idProperty = ( schema.properties || {})[schema['exegesis-plugin-jsonschema-id'] || 'id'] || { type: 'integer' };
            this.processSchema( Object.assign({}, context, {
                urlPath: `${context.urlPath || '' }/{pathParam${( context.pathParams || [] ).length}}`,
                pathParams: ( context.pathParams || [] ).concat( [ {
                    'in': 'path',
                    name: `pathParam${( context.pathParams || [] ).length}`,
                    schema: idProperty,
                    required: true,
                    description: `ID for ${schema.title || 'item' }`
                } ] )
            }), schema, false, !schema.$id );
        } else {
            if( !skipPath ) {
                if( schema.$id ) {
                    this.addObjectDefinition( context, schema );
                }
                this.addObjectPath( context, schema );
            }
            if( schema.properties ) {
                Object.keys( schema.properties ).forEach( name => {
                    let definition = schema.properties[name];
                    if( definition.type === 'array' && definition.items ) {
                        this.processSchema( Object.assign({}, context, {
                            urlPath: `${context.urlPath}/${name}`
                        }), definition.items, true, !definition.items.$id );
                    } else if( definition.type === 'object' || definition.properties ) {
                        this.processSchema( Object.assign({}, context, {
                            urlPath: `${context.urlPath}/${name}`
                        }), definition, false, !definition.$id );
                    }
                });
            }
            if( schema.operations ) {
                Object.keys( schema.operations ).forEach( name => {
                    let definition = schema.operations[name];
                    this.addOperationPath( Object.assign({}, context, {
                        urlPath: `${context.urlPath}/${name}`
                    }), schema, definition );
                });
            }
        }
    }
}

module.exports = function pathGenerator( apiDoc, controller, schema, baseUrl ) {
    if( !schema || ( schema.type !== undefined && schema.type !== 'object' ) || !schema.properties ) {
        throw new Error( 'Root schema must have object type and have properties' );
    }
    var generator = new JSONSchemaPathGenerator( apiDoc, controller, baseUrl );
    return generator.processSchema({ urlPath: '', pathParams: [] }, schema, false, true );
};


src/controller.js
'use strict';

const controllerBaseName = 'exegesis-plugin-jsonschema-';
var count = 0;

module.exports = class JSONSchemaController {
    constructor() {
        this._controllers = null;
        this._name = controllerBaseName + count++;
    }

    get name() {
        return this._name;
    }

    set controllers( value ) {
        if( this._controllers ) {
            throw new Error( 'Controller was already initialized' );
        }
        this._controllers = value;
    }

    resolvePath( reqContext ) {
        reqContext.requestObjectPath = ( reqContext.api.pathItemObject['x-exegesis-jsonschema-pathtemplate'] || '' ).split( '/' ).map( item => {
            var match = item.match( /^\{(.*)\}$/ );
            if( match ) {
                return reqContext.params.path[match[1]];
            }
            return item;
        });
    }

    getController( reqContext, operation ) {
        if( !this._controllers ) {
            throw reqContext.makeError( 500, 'JSONSchemaController controllers have not been initialized' );
        }

        var controllerName = reqContext.api.operationObject['x-exegesis-jsonschema-controller'] ||
            reqContext.api.pathItemObject['x-exegesis-jsonschema-controller'];

        var controller = this._controllers[controllerName];

        if( !controller ) {
            throw reqContext.makeError( 400, `${controllerName} controller not implemented` );
        }

        if( typeof( controller[operation] ) !== 'function' ) {
            throw reqContext.makeError( 400, `${operation} not implemented in ${controllerName} controller` );
        }

        if( typeof( controller.formatResponse ) !== 'function' ) {
            throw reqContext.makeError( 400, `formatResponse not implemented in ${controllerName} controller` );
        }

        return controller;
    }

    getBlankObject( reqContext ) {
        return JSON.parse( reqContext.api.operationObject['x-exegesis-jsonschema-blankobject'] ||
                           reqContext.api.pathItemObject['x-exegesis-jsonschema-blankobject'] || '{}' );
    }

    async getItem( reqContext ) {
        var controller = this.getController( reqContext, 'get' );
        this.resolvePath( reqContext );
        var item = await controller.get( reqContext );

        if( !item ) {
            throw reqContext.makeError( 404, 'Not found' );
        }

        return controller.formatResponse( reqContext, item );
    }

    async patchItem( reqContext ) {
        var controller = this.getController( reqContext, 'update' );
        this.resolvePath( reqContext );
        var item = await controller.update( reqContext );

        if( !item ) {
            throw reqContext.makeError( 404, 'Not found' );
        }

        return await controller.formatResponse( reqContext, item );
    }

    async putItem( reqContext ) {
        var controller = this.getController( reqContext, 'update' );
        this.resolvePath( reqContext );

        reqContext.requestBody = Object.assign(
            this.getBlankObject( reqContext ),
            reqContext.requestBody );
        var item = await controller.update( reqContext );

        if( !item ) {
            throw reqContext.makeError( 404, 'Not found' );
        }

        return await controller.formatResponse( reqContext, item );
    }

    async deleteItem( reqContext ) {
        var controller = this.getController( reqContext, 'remove' );
        this.resolvePath( reqContext );

        var item = await controller.remove( reqContext );

        if( !item ) {
            throw reqContext.makeError( 404, 'Not found' );
        }

        reqContext.res.setStatus( 204 );
    }

    async getItems( reqContext ) {
        var controller = this.getController( reqContext, 'find' );
        this.resolvePath( reqContext );
        var items = await controller.find( reqContext );

        if( !items ) {
            throw reqContext.makeError( 404, 'Not found' );
        }

        if( !Array.isArray( items ) ) {
            throw reqContext.makeError( 500, 'find must return an array' );
        }
        items = items.map( async( item ) => { return controller.formatResponse( reqContext, item ); });
        return Promise.all( items );
    }

    async postItems( reqContext ) {
        var controller = this.getController( reqContext, 'create' );
        this.resolvePath( reqContext );

        reqContext.requestBody = Object.assign(
            this.getBlankObject( reqContext ),
            reqContext.requestBody );

        var item = await controller.create( reqContext );

        if( !item ) {
            throw reqContext.makeError( 404, 'Not found' );
        }

        return await controller.formatResponse( reqContext, item );
    }

    async callOperation( reqContext ) {
        var controller = this.getController( reqContext, 'callOperation' );
        this.resolvePath( reqContext );

        reqContext.requestBody = Object.assign(
            this.getBlankObject( reqContext ),
            reqContext.requestBody );

        var item = await controller.callOperation( reqContext );

        if( !item ) {
            throw reqContext.makeError( 404, 'Not found' );
        }

        return await controller.formatResponse( reqContext, item );
    }

};
